test/rematch2.jl

```diff
@@ -629,15 +709,36 @@ end
         end
     end
 
+    if VERSION >= v"1.8"
+        @testset "type constraints on the input are observed" begin
+            let line = (@__LINE__) + 7
+                @test_warn(
+                    "$file:$line: Case 4: `_ =>` is not reachable.",
+                    @eval @match2 identity(BoolPair(true, false))::BoolPair begin
+                        BoolPair(true, _)       => 1
+                        BoolPair(_, true)       => 2
+                        BoolPair(false, false)  => 3
+                        _                       => 4 # unreachable
+                    end
+                    )
+            end
+        end
+    end
```

Comment:
Here

Reply:
ah, thanks! so we know that 4 is unreachable because of the booelan checks above - cool!!